File: drive/src/main/java/de/mein/drive/bash/BashTools.java
package de.mein.drive.bash;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.List;
import java.util.Map;
import java.util.Set;

import de.mein.Lok;
import de.mein.auth.file.AFile;

/**
 * Created by xor on 10/28/16.
 */
@SuppressWarnings("Duplicates")
public abstract class BashTools {

    public static final boolean isWindows = System.getProperty("os.name").toLowerCase().startsWith("windows");
    private static BashToolsImpl instance;

    public static BashToolsImpl getInstance() {
        return instance;
    }

    public static void setInstance(BashToolsImpl instance) {
        BashTools.instance = instance;
    }

    public static void init() {
        if (instance == null)
            if (isWindows) {
                instance = new BashToolsWindows();
            } else {
                instance = new BashToolsUnix();
            }
    }

    public static void setBinPath(String binPath) {
        instance.setBinPath(binPath);
    }

    public static FsBashDetails getFsBashDetails(AFile file) throws IOException, InterruptedException {
        return instance.getFsBashDetails(file);
    }

    public static void rmRf(AFile directory) throws IOException {
        instance.rmRf(directory);
    }

    public static void rmRf(File directory) throws IOException {
        instance.rmRf(AFile.instance(directory));
    }

    public static List<AFile<?>> stuffModifiedAfter(AFile referenceFile, AFile directory, AFile pruneDir) throws IOException, BashToolsException {
        return instance.stuffModifiedAfter(referenceFile, directory, pruneDir);
    }

    public static AutoKlausIterator<AFile<?>> find(AFile directory, AFile pruneDir) throws IOException {
        return instance.find(directory, pruneDir);
    }

    public static AutoKlausIterator<AFile<?>> stuffModifiedAfter(AFile originalFile, AFile pruneDir, long timeStamp) throws IOException, InterruptedException {
        return instance.stuffModifiedAfter(originalFile, pruneDir, timeStamp);
    }

    public static AutoKlausIterator<AFile> inputStreamToFileIterator(InputStream inputStream) {
        BufferedIterator bufferedReader = new BufferedIterator.BufferedFileIterator(new InputStreamReader(inputStream));
        return bufferedReader;
    }

    public static AutoKlausIterator<String> inputStreamToIterator(InputStream inputStream) {
        BufferedIterator bufferedReader = new BufferedIterator.BufferedStringIterator(new InputStreamReader(inputStream));
        return bufferedReader;
    }

    public static void mkdir(AFile dir) throws IOException {
        int i = 0;
        while (!dir.exists()) {
            dir.mkdirs();
            Lok.debug("BashTools.mkdir(" + i + ") for " + dir.getAbsolutePath());
        }
        //instance.mkdir(dir);
    }

    public static File[] lsD(String path) {
        File root = new File(path);
        if (root.exists()) {
            File[] dirs = root.listFiles((file, s) -> file.isDirectory());
            return dirs;
        }
        return new File[0];
    }


    public static boolean isSymLink(AFile f) {
        return instance.isSymLink(f);
    }

    public static Map<String, FsBashDetails> getContentFsBashDetails(AFile file) {
        return instance.getContentFsBashDetails(file);
    }

    public static void lnS(AFile file, String target) {
        Lok.debug("creating symlink: '" + file.getAbsolutePath() + "' -> '" + target + "'");
        instance.lnS(file, target);
    }

    public static void setCreationDate(AFile target, Long created) {
        if (created != null)
            instance.setCreationDate(target, created);
    }
}

File: filesync/src/main/kotlin/de/mel/filesync/bash/BashTools.kt
package de.mel.filesync.bash

import de.mel.auth.file.AbstractFile
import de.mel.auth.file.IFile
import de.mel.auth.file.StandardFile
import de.mel.filesync.bash.BufferedIterator.BufferedFileIterator
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.attribute.FileTime


abstract class BashTools<A : IFile> {
    abstract fun getFsBashDetails(file: A): FsBashDetails?

    companion object {
        var binPath: String? = null
        val isWindows = System.getProperty("os.name").toLowerCase().startsWith("windows")
        lateinit var implementation: BashTools<IFile>
        fun <F : IFile> getFsBashDetails(file: F): FsBashDetails? = implementation.getFsBashDetails(file)

        fun init() {
            if (implementation == null) if (BashTools.isWindows) {
                implementation = BashToolsWindows() as BashTools<IFile>
            } else {
                implementation = BashToolsUnix() as BashTools<IFile>
            }
        }

        fun inputStreamToFileIterator(inputStream: InputStream): AutoKlausIterator<AbstractFile<*>> {
            return BufferedFileIterator(InputStreamReader(inputStream))
        }

        fun rmRf(file: IFile) = implementation.rmRf(file)
        fun rmRf(file: File) = implementation.rmRf(AbstractFile.instance(file))

        fun find(root: IFile, pruneDir: IFile): AutoKlausIterator<IFile> = implementation.find(root, pruneDir)
        fun isSymLink(file: IFile) = implementation.isSymLink(file)

        fun getContentFsBashDetails(dir: IFile) = implementation.getContentFsBashDetails(dir)
        fun lnS(symbolic: IFile, target: String) = implementation.lnS(symbolic, target)
        fun stuffModifiedAfter(directory: IFile, pruneDir: IFile, timeStamp: Long): AutoKlausIterator<IFile> = implementation.stuffModifiedAfter(directory, pruneDir, timeStamp)
        fun stuffModifiedAfter(referenceFile: IFile, directory: IFile, pruneDir: IFile): List<IFile> = implementation.stuffModifiedAfter(referenceFile, directory, pruneDir)
        fun setCreationDate(target: IFile, created: Long) = implementation.setCreationDate(target = target, created = created)
    }

    open fun isSymLink(f: A): Boolean = Files.isSymbolicLink(Paths.get(File(f.absolutePath).toURI()))

    abstract fun lnS(file: A, target: String)
    open fun setBinPath(binPath: String) {
        // Override if needed
    }

    open fun setCreationDate(target: A, created: Long) {
        val path = Paths.get(File(target.absolutePath).toURI())
        Files.setAttribute(path, "creationTime", FileTime.fromMillis(created))
    }

    // todo this can be replaces by calling constructor
    @Deprecated(level = DeprecationLevel.WARNING, message = "should be replaced by calling constructor directly", replaceWith = ReplaceWith("FsBashDetails(created, modified, iNode, symLink, symLinkTarget, name)", "de.mel.filesync.bash.FsBashDetails"))
    open fun createFsBashDetails(created: Long?, modified: Long, iNode: Long, symLink: Boolean, symLinkTarget: String?, name: String): FsBashDetails {
        return FsBashDetails(created, modified, iNode, symLink, symLinkTarget, name)
    }

    abstract fun getContentFsBashDetails(directory: A): MutableMap<String, FsBashDetails>
    @Throws(exceptionClasses = [IOException::class])
    abstract fun rmRf(directory: A)

    //todo replace return type with AutoKlausIterator
    @Throws(exceptionClasses = [IOException::class, de.mel.filesync.bash.BashToolsException::class])
    abstract fun stuffModifiedAfter(referenceFile: A, directory: A, pruneDir: A): List<A>

    @Throws(exceptionClasses = [IOException::class])
    abstract fun find(directory: A, pruneDir: A): AutoKlausIterator<A>

    @Throws(exceptionClasses = [IOException::class, java.lang.InterruptedException::class])
    abstract fun stuffModifiedAfter(directory: A, pruneDir: A, timeStamp: Long): AutoKlausIterator<A>
}